src/lib.rs

```diff
@@ -229,11 +284,14 @@ impl AllocList {
 
     /// Updates the size associated with a base pointer (perfomed on a realloc).
     /// This must not be called with an inner pointer.
-    fn update(&mut self, ptr: usize, size: usize) {
+    fn update(&mut self, ptr: usize, new_ptr: usize, size: usize) {
         for block in self.iter_mut() {
             if let Block::Entry(base_ptr, _) = *block {
                 if ptr == base_ptr.get() {
-                    *block = Block::Entry(base_ptr, size);
+                    *block = Block::Entry(
+                        unsafe { core::num::NonZeroUsize::new_unchecked(new_ptr) },
```

Comment:
I feel like a `debug_assert_ne!(new_ptr, 0)` or similar might be a good idea, just to check that nobody ever violates the `NonZeroUsize` property (because if they do, bad things can happen).

Reply:
I think you're right and there's probably not much harm in doing so. In most cases these calls are made by the compiler, so you'd hope they followed their own rules. If we add assertions, I think these checks should probably happen in the actual `alloc`, `dealloc`, `realloc` calls instead. 

Rust makes other assumptions about allocation which would result in UB which are perfectly legal in system `malloc` / `free` / `realloc`. For example, if `libc::realloc` is called with `size = 0`, it acts as `libc::free`. In rust this is UB. I'm guessing they are more restrictive for portability with other allocator implementations.

Reply:
In general (and it's hard to make a hard rule for this), I think it's not a bad idea to have a `debug_assert` before an `unsafe` block to check that whatever condition is expected to hold really holds. In some senses, this is often as useful as documentation as it is an actual check. In other words, I don't mind having "duplicate" `debug_assert`'s around, because it can help give us confidence about individual `unsafe` blocks even if we don't understand the full system context.

Reply:
> In most cases these calls are made by the compiler

So an assertion could serve as  a check of our understanding of the compiler ;)